File: src/sha256pow/search/result.py
from __future__ import annotations

from dataclasses import dataclass
from typing import Optional


@dataclass
class SearchResult:
    nonce: int
    digest_hex: str
    tries: int
    elapsed_sec: float
    hashes_per_sec: float
    double_hash: bool
    k: int
    nonce_size: Optional[int] = None
    backend: str = "unknown"
    workers: int = 1
    prefix_hex: Optional[str] = None

    @property
    def mhashes_per_sec(self) -> float:
        return self.hashes_per_sec / 1e6

    def format_block(self) -> str:
        lines = [
            "[Search Result]",
            "-----------------------------",
            f"backend         : {self.backend}",
            f"workers         : {self.workers}",
            f"nonce           : {self.nonce}",
            f"digest (hex)    : {self.digest_hex}",
            f"tries           : {self.tries:,}",
            f"elapsed (sec)   : {self.elapsed_sec:.4f}",
            f"throughput      : {self.hashes_per_sec:,.2f} H/s ({self.mhashes_per_sec:.2f} MH/s)",
            f"double hash     : {self.double_hash}",
            f"target bits (k) : {self.k}",
        ]
        if self.nonce_size is not None:
            lines.append(f"nonce size      : {self.nonce_size} bytes")
        if self.prefix_hex is not None:
            lines.append(f"prefix (hex)    : {self.prefix_hex}")
        return "\n".join(lines)

    def __str__(self) -> str:
        return "\n" + self.format_block() + "\n"

File: native/rust/src/result.rs
use crate::progress::format_with_commas;

#[derive(Debug, Clone)]
pub struct SearchResult {
    pub nonce: u64,
    pub digest_hex: String,
    pub tries: u64,
    pub elapsed_sec: f64,
    pub hashes_per_sec: f64,
    pub double_hash: bool,
    pub k: u32,
    pub nonce_size: usize,
    pub backend: &'static str,
    pub workers: usize,
    pub prefix_hex: Option<String>,
}

impl std::fmt::Display for SearchResult {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        writeln!(f)?;
        writeln!(f, "[Search Result]")?;
        writeln!(f, "-----------------------------")?;
        writeln!(f, "backend         : {}", self.backend)?;
        writeln!(f, "workers         : {}", self.workers)?;
        writeln!(f, "nonce           : {}", self.nonce)?;
        writeln!(f, "digest (hex)    : {}", self.digest_hex)?;
        writeln!(f, "tries           : {}", format_with_commas(self.tries))?;
        writeln!(f, "elapsed (sec)   : {:.4}", self.elapsed_sec)?;
        writeln!(
            f,
            "throughput      : {:.2} H/s ({:.2} MH/s)",
            self.hashes_per_sec,
            self.hashes_per_sec / 1e6
        )?;
        writeln!(f, "double hash     : {}", self.double_hash)?;
        writeln!(f, "target bits (k) : {}", self.k)?;
        writeln!(f, "nonce size      : {} bytes", self.nonce_size)?;
        if let Some(prefix_hex) = &self.prefix_hex {
            writeln!(f, "prefix (hex)    : {}", prefix_hex)?;
        }
        Ok(())
    }
}
